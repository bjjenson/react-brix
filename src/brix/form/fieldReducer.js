export const actionTypes = {
  insertField: 'insertField',
  removeField: 'removeField',
  reset: 'reset',
  touched: 'touched',
  updateValue: 'updateValue',
  validationResult: 'validationResult',
}

export const actions = {
  insertField: (fieldName, fieldState) => ({ type: actionTypes.insertField, fieldName, payload: fieldState }),
  removeField: fieldName => ({ type: actionTypes.removeField, fieldName }),
  reset: (state) => ({ type: actionTypes.reset, payload: state }),
  touched: (fieldName) => ({ type: actionTypes.touched, fieldName }),
  updateValue: (fieldName, value) => ({ type: actionTypes.updateValue, fieldName, payload: value }),
  validationResult: (fieldName, error, helperText) => ({ type: actionTypes.validationResult, fieldName, payload: { error, helperText } }),
}

export const initState = state => {
  return state
}

const current = 'current'

export const fieldReducer = (state, { type, fieldName, payload }) => {
  const handlers = {
    [actionTypes.updateValue]: value =>
      state.setIn([fieldName, current, 'value'], value)
        .setIn([fieldName, current, 'pristine'], value == state.getIn([fieldName, 'initial', 'value'])),

    [actionTypes.touched]: () =>
      state.setIn([fieldName, current, 'touched'], true),

    [actionTypes.validationResult]: ({ error, helperText }) =>
      state.setIn([fieldName, current, 'error'], error)
        .setIn([fieldName, current, 'helperText'], helperText),

    [actionTypes.insertField]: fieldState =>
      state.set(fieldName, fieldState),

    [actionTypes.removeField]: () =>
      state.delete(fieldName),

    [actionTypes.reset]: state => state,
  }

  const handler = handlers[type]
  if (handler) {
    return handler(payload)
  }
  return state
}
