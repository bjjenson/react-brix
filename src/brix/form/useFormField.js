import { actions } from './fieldReducer'
/**
 * @param  fieldArgs { import("../form").IFormFieldArgs}
 * @returns {import("../form").IFormField}
 */
export const useFormField = (state, dispatch, fieldArgs = {}) => {
  const requiredMessage = fieldArgs.requiredMessage || 'Required'

  const tryValidate = (value, touched) => {
    let result
    if (value && fieldArgs.validate) {
      result = fieldArgs.validate(value, fieldArgs.name, state.get('getAllValues'))
    }

    if (!Boolean(result) && !fieldArgs.optional && touched && value === '') {
      result = requiredMessage
    }

    if (touched) {
      dispatch(actions.validationResult(fieldArgs.name, Boolean(result), result))
    }
    return !Boolean(result)
  }

  const setValue = v => {
    dispatch(actions.updateValue(fieldArgs.name, v))
    tryValidate(v, state.getIn(['current', 'touched']))
  }

  const onChange = ({ target }) => {
    const value = fieldArgs.valueFromTarget ? fieldArgs.valueFromTarget(target) : target.value
    const coercedValue = fieldArgs.normalize ? fieldArgs.normalize(value) : value
    setValue(coercedValue)
  }

  const onBlur = () => {
    dispatch(actions.touched(fieldArgs.name))
    tryValidate(state.getIn(['current', 'value']), true)
  }

  const setValidationResult = result => {
    dispatch(actions.validationResult(fieldArgs.name, true, result))
  }

  const validate = () => {
    return tryValidate(state.getIn(['current', 'value']), true)
  }

  return {
    props: {
      error: state.getIn(['current', 'error']),
      helperText: state.getIn(['current', 'helperText']),
      label: state.getIn(['initial', 'label']),
      value: state.getIn(['current', 'value']),
      // handlers
      onBlur,
      onChange,
    },
    setValidationResult,
    setValue,
    validate,
  }
}
