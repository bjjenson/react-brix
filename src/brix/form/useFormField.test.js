import { fromJS } from 'immutable'
import { useFormField } from './useFormField'
import { actionTypes } from './fieldReducer'

jest.mock('react')

const dispatch = jest.fn()
let value, initialArgs, state
beforeEach(() => {
  value = 'the value'
  initialArgs = {
    name: 'fieldName',
    optional: false,
    error: 'on',
    helperText: 'the help',
    label: 'the label',
  }

  state = fromJS({
    initial: {
      label: initialArgs.label,
      type: 'text',
      value,
      optional: false,
    },
    current: {
      value,
      error: false,
      pristine: true,
      touched: false,
      helperText: initialArgs.helperText,
    },
    getAllValues: 'getAllValuesFunc',
  })
})

test('returns all props needed', () => {
  expect(useFormField(state, dispatch, initialArgs)).toMatchSnapshot()
})

describe('onChange', () => {
  let event
  beforeEach(() => {
    event = {
      target: {
        value: 'updated',
      },
    }
  })

  test('dispatches change action', () => {
    const { props: { onChange } } = useFormField(state, dispatch, initialArgs)
    onChange(event)

    expect(dispatch.mock.calls[0]).toMatchSnapshot()
  })

  test('uses valueFromTarget to coerce value if provided', () => {
    const valueFromTarget = jest.fn()
    valueFromTarget.mockReturnValue('updatedFromTarget')
    const { props: { onChange } } = useFormField(state, dispatch, { ...initialArgs, valueFromTarget })
    onChange(event)

    expect(dispatch.mock.calls[0]).toMatchSnapshot()
    expect(valueFromTarget).toHaveBeenCalledWith(event.target)
  })

  test('coerces value using normalize', () => {
    const normalize = jest.fn()
    normalize.mockReturnValue('normalized value')
    const { props: { onChange } } = useFormField(state, dispatch, { ...initialArgs, normalize })
    onChange(event)

    expect(dispatch.mock.calls[0]).toMatchSnapshot()
    expect(normalize).toHaveBeenCalledWith(event.target.value)
  })
})

describe('validate', () => {
  const validate = jest.fn()
  let event
  beforeEach(() => {
    initialArgs = { ...initialArgs, validate }
    event = {
      target: {
        value: '1',
      },
    }
    state = fromJS({
      current: {
        value,
        error: false,
        pristine: false,
        touched: true,
        helperText: initialArgs.helperText,
      },
      getAllValues: 'getAllValuesFunc',
    })
  })

  test('calls validate on change', () => {
    const { props: { onChange } } = useFormField(state, dispatch, initialArgs)
    onChange(event)

    expect(validate.mock.calls[0]).toMatchSnapshot()
  })

  test('does not set error if not "touched"', () => {
    state = fromJS({
      current: {
        value,
        error: false,
        pristine: false,
        touched: false,
        helperText: initialArgs.helperText,
      },
    })
    validate.mockReturnValue('Error here')

    const { props: { onChange } } = useFormField(state, dispatch, initialArgs)
    onChange(event)

    expect(dispatch).not.toHaveBeenCalledWith({
      type: actionTypes.validationResult,
      payload: {
        error: true,
        helperText: expect.any(String),
      },
    })

  })

  test('sets validation result if "touched"', () => {
    validate.mockReturnValue('Error here')

    const { props: { onChange } } = useFormField(state, dispatch, initialArgs)
    onChange(event)

    expect(dispatch).toHaveBeenCalledWith({
      type: actionTypes.validationResult,
      fieldName: 'fieldName',
      payload: {
        error: true,
        helperText: expect.any(String),
      },
    })

  })

  test('validates onBlur', () => {
    validate.mockReturnValue('Error here')

    const { props: { onBlur } } = useFormField(state, dispatch, initialArgs)
    onBlur()

    expect(validate.mock.calls[0]).toMatchSnapshot()
  })

  test('sets required if empty on blur', () => {
    state = fromJS({
      current: {
        value: '',
        error: false,
        pristine: false,
        touched: true,
        helperText: initialArgs.helperText,
      },
    })

    const { props: { onBlur } } = useFormField(state, dispatch, initialArgs)
    onBlur()

    expect(dispatch).toHaveBeenCalledWith({
      type: actionTypes.validationResult,
      fieldName: 'fieldName',
      payload: {
        error: true,
        helperText: expect.any(String),
      },
    })
  })

  test('validate directly', () => {
    state = fromJS({
      current: {
        value: '',
        error: false,
        pristine: false,
        touched: false,
        helperText: initialArgs.helperText,
      },
    })

    const actual = useFormField(state, dispatch, initialArgs)

    const isValid = actual.validate()

    expect(dispatch).toHaveBeenCalledWith({
      type: actionTypes.validationResult,
      fieldName: 'fieldName',
      payload: {
        error: true,
        helperText: 'Required',
      },
    })
    expect(isValid).toBeFalsy()
  })

  test('validate directly returns true if valid', () => {
    validate.mockReturnValue('')
    const actual = useFormField(state, dispatch, initialArgs)

    const isValid = actual.validate()

    expect(isValid).toBeTruthy()
  })

})

test('setValidationResult externally', () => {
  const { setValidationResult } = useFormField(state, dispatch, initialArgs)

  setValidationResult('i am error')
  expect(dispatch).toHaveBeenCalledWith({
    type: actionTypes.validationResult,
    fieldName: 'fieldName',
    payload: {
      error: true,
      helperText: 'i am error',
    },
  })
})

test('sets touched onBlur', () => {
  const { props: { onBlur } } = useFormField(state, dispatch, initialArgs)
  onBlur()

  expect(dispatch).toHaveBeenCalledWith({
    fieldName: 'fieldName',
    type: actionTypes.touched,
  })
})

test('setValue sets value externally', () => {
  const { setValue } = useFormField(state, dispatch, initialArgs)

  setValue('i am value')
  expect(dispatch).toHaveBeenCalledWith({
    type: actionTypes.updateValue,
    fieldName: 'fieldName',
    payload: 'i am value',
  })
})

