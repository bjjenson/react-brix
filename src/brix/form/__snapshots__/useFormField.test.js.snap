// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adds optional to label if not required 1`] = `
Object {
  "props": Object {
    "error": false,
    "helperText": "the help",
    "label": "the label (optional)",
    "onBlur": [Function],
    "onChange": [Function],
    "value": "the value",
  },
  "setValidationResult": [Function],
  "setValue": [Function],
  "validate": [Function],
}
`;

exports[`onChange coerces value using normalize 1`] = `
Array [
  Object {
    "fieldName": "fieldName",
    "payload": "normalized value",
    "type": "updateValue",
  },
]
`;

exports[`onChange dispatches change action 1`] = `
Array [
  Object {
    "fieldName": "fieldName",
    "payload": "updated",
    "type": "updateValue",
  },
]
`;

exports[`onChange uses valueFromTarget to coerce value if provided 1`] = `
Array [
  Object {
    "fieldName": "fieldName",
    "payload": "updatedFromTarget",
    "type": "updateValue",
  },
]
`;

exports[`returns all props needed 1`] = `
Object {
  "props": Object {
    "error": false,
    "helperText": "the help",
    "label": "the label",
    "onBlur": [Function],
    "onChange": [Function],
    "value": "the value",
  },
  "setValidationResult": [Function],
  "setValue": [Function],
  "validate": [Function],
}
`;
